### PR TITLE
Some more allocation reduction in SocketsHttpHandler

### PR DESCRIPTION
**Pull request by stephentoub**
- Pool SocketAsyncEventArgs used for establishing connections.  Today we create a new SocketAsyncEventArgs for each new connection, and it's used only for the ConnectAsync.  This establishes a bounded pool of them, such that they can be reused.  To do this, I moved the nested ConcurrentQueue.Segment type into a partial file so that we could use that directly, as it supports bounding (we can use it as a public type if/when https://github.com/dotnet/corefx/issues/24365 is addressed). FYI, @kouvel.

- Zero pool lifetimes. If the PooledConnectionIdleTimeout is zero or if the PooledConnectionLifetime is zero, then no connection will ever actually be usuable from the pool.  In such a situation we can a) avoid creating the cleaning timer, and b) dispose of the connection as it's being returned to the pool, rather than waiting for it to be scavenged.

- CookieContainer string allocation.  Its ctor is allocating a string that can instead be constructed once and used across all CookieContainers.

cc: @geoffkizer

### CODE REVIEW

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectHelper.cs [hunk 1]
```diff
@@ -34,86 +40,100 @@ public CertificateCallbackMapper(Func<HttpRequestMessage, X509Certificate2, X509
...
-                // to cancel it if needed.
-                using (var saea = new BuilderAndCancellationTokenSocketAsyncEventArgs(cancellationToken))
-                {
-                    // Configure which server to which to connect.
-                    saea.RemoteEndPoint = IPAddress.TryParse(host, out IPAddress address) ?
-                        (EndPoint)new IPEndPoint(address, port) :
-                        new DnsEndPoint(host, port);
+                saea = new ConnectEventArgs();
+            }
+            saea.Initialize(cancellationToken);
```
Comment (pjanotti) [inline src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectHelper.cs]: Q about style: I thought that we usually added a blank line after a block, no?

**Comment by kouvel** [comment]
Queue segment changes look good to me

**Comment by stephentoub** [comment]
@dotnet-bot test this please

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectHelper.cs (hunk 1 above)
Comment (stephentoub) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectHelper.cs]: Yeah, we generally do, and I probably should have here.

**Comment by stephentoub** [comment]
@dotnet-bot test Outerloop Windows x64 Debug Build please

**Pull request merged**